include/rmx/rmx.hpp

```diff
@@ -0,0 +1,85 @@
+#pragma once
+#include <functional>
+#include <mutex>
+#include <optional>
+#include <type_traits>
+
+namespace rmx {
+
+//! An RAII-style guard wrapping a reference to some type protected by a mutex
+//!
+//! Acquire a `MutexGuard` by locking a `Mutex`.
+template<typename ValueT, typename MutexImplT>
+class MutexGuard
+{
+  public:
+    explicit MutexGuard(ValueT& value_ref, std::unique_lock<MutexImplT>&& lock) noexcept :
+        m_lock(std::move(lock)), m_ref(value_ref)
+    {
+    }
+
+    //! Access the underlying value by reference
+    //!
+    //! @warning It is incorrect to store the reference returned by this operator.
+    [[nodiscard]] ValueT& operator*() noexcept { return m_ref; }
+    [[nodiscard]] const ValueT& operator*() const noexcept { return m_ref; }
+
+    //! Access the underlying value by pointer
+    //!
+    //! @warning It is incorrect to store the pointer returned by this operator.
+    [[nodiscard]] ValueT* operator->() noexcept { return &m_ref.get(); }
+    [[nodiscard]] const ValueT* operator->() const noexcept { return &m_ref.get(); }
+
+    //! Provide access to the underlying std::unique_lock to facilitate use with
+    //! std::condition_variable::wait()
+    //!
+    //! @warning It is incorrect to access the wrapped value if the inner lock has been manually
+    //! unlocked. Don't do that.
+    [[nodiscard]] std::unique_lock<MutexImplT>& inner() noexcept { return m_lock; }
```

Comment:
Exposing this enables the user to do dangerous and dumb things. But it seems it's also necessary if we want interopt with `std::condition_variable::wait()`

Reply:
Another option could be to implement `lock()` and `unlock()` directly on this type; then it would interop with `condition_variable_any` directly. I've never used it; but it's the general form of `condition_variable` that just maybe isn't quite as performant.

But that could work along with your comment about unlocking it - rather than unlocking the root object itself, unlocking the guard is probably eaiser. Then your value access assertions could just be `assert(m_lock.owns_lock());` (see `#include <cassert>`)